### PR TITLE
Elastic search completion suggester for type ahead

### PR DESCRIPTION
**Pull request by camelliazhang**
In this check in, 
1, Updated ES custom analyzers with whitespace tokenizer and word delimiter filter
For the fields of dataset name, metric name, flow name and job name
2, Adopted Elastic search completion suggester which is optimized for speed. This also eliminates the access to mySQL in DAO for auto complete. In addition, this also lays the foundation for future scoring of type aheads. For all datasets/metrics/flows

### CODE REVIEW

File: web/app/dao/SearchDAO.java [hunk 1]
```diff
@@ -183,34 +164,94 @@
...
 	public static List<String> getAutoCompleteListDataset(String input, int limit) {
-		String pattern = isNotBlank(input) ? input + "%" : "%";
-		if (limit <= 0) {
-			return getJdbcTemplate().queryForList(GET_DATASET_AUTO_COMPLETE_LIST_NO_LIMIT, String.class, pattern);
-		}
-		return getJdbcTemplate().queryForList(GET_DATASET_AUTO_COMPLETE_LIST, String.class, pattern, limit);
+		String elasticSearchTypeURLKey = "elasticsearch.dataset.url";
+		String fieldName = "name_suggest";
+		return getAutoCompleteListbyES(elasticSearchTypeURLKey,fieldName,input,limit);
+	}
```
Comment (mars-lan) [inline web/app/dao/SearchDAO.java]: Insert new line after for readability

File: web/app/dao/SearchDAO.java [hunk 2]
```diff
@@ -183,34 +164,94 @@
...
+			return completionSuggestionList;
+		}
+
+		Iterator<JsonNode> arrayIterator = whSuggestNode.elements();
+		if (arrayIterator == null) return completionSuggestionList;
+
+		while (arrayIterator.hasNext())
+		{
+			JsonNode node = arrayIterator.next();
+			if (node.isContainerNode() && node.has("options"))
```
Comment (mars-lan) [inline web/app/dao/SearchDAO.java]: Let's get rid of the nested if statements with early terminations, i.e. something like if (!node.isContainerNode() || !node.has("options")) continue;  to improve readability.

File: web/app/dao/SearchDAO.java [hunk 3]
```diff
@@ -183,34 +164,94 @@
...
-		if (limit <= 0) {
-			return getJdbcTemplate().queryForList(GET_METRIC_AUTO_COMPLETE_LIST_NO_LIMIT, String.class, pattern);
+	public static List<String> getAutoCompleteListbyES(String elasticSearchTypeURLKey, String fieldName, String input, int limit) {
+		// use elastic search completion suggester, ES will validate the input and limit
+		List<String> completionSuggestionList = new ArrayList<String>();
+		JsonNode responseNode = null;
+		ObjectNode keywordNode = null;
+
+		try
+		{
```
Comment (mars-lan) [inline web/app/dao/SearchDAO.java]: Let's try keep the styling more consistent here. Use K&R (https://en.wikipedia.org/wiki/Indent_style#K.26R) like the rest of the code.

**Comment by camelliazhang** [comment]
@mars-lan , can you review it again when you get a chance? Thanks, Na

**Comment by camelliazhang** [comment]
@mars-lan , reformat a few functions. Please let me know if there are any further questions! Thanks, Na

**Pull request merged**